File: 2020/02/02.py
from hashlib import new
from typing import Generator
from parse import parse

def preprocessing(input: str) -> Generator[tuple[int, int, str, str], None, None]: 
    return (parse("{:d}-{:d} {}: {}", pw) for pw in input.splitlines())

def solver(passwords: Generator[tuple[int, int, str, str], None, None]):
    old_rule: int = 0
    new_rule: int = 0
    for a, b, w, pw in passwords:
        if a <= pw.count(w) <= b: old_rule += 1
        if (pw[a-1] == w) ^ (pw[b-1] == w): new_rule += 1
    yield old_rule
    yield new_rule

File: 2020/02/02.rs
aoc2020::main!();

fn generator(input: &str) -> Vec<(usize, usize, char, &str)> {
    input
    .lines()
    .map(|l| l
        .split(['-',' ',':'])
        .collect_tuple()
        .unwrap())
    .map(|(a,b,c,_,d)| (a.parse().unwrap(), b.parse().unwrap(), c.chars().nth(0).unwrap(), d))
    .collect_vec()
}

fn part_1(input: Vec<(usize, usize, char, &str)>) -> usize {
    input
    .iter()
    .filter(|(a,b,c,d)| 
        (a..=b)
        .contains(&& d.chars().filter(|ch| ch == c)
        .count()))
    .count()
}

fn part_2(input: Vec<(usize, usize, char, &str)>) -> usize {
    input
    .iter()
    .filter(|&(a,b,c,d)| (d.chars().nth(a - 1).unwrap() == *c) ^ (d.chars().nth(b - 1).unwrap() == *c))
    .count()
}